src/parser.rs

```diff
@@ -1,81 +1,35 @@
-/// deserialized version of the spec file
-#[derive(Debug, Serialize, Deserialize)]
-pub struct AsyncApi {
-    pub asyncapi: String,
-    pub info: Info,
-    pub servers: HashMap<String, Server>,
-    pub channels: HashMap<String, Channel>,
-}
-
-#[derive(Debug, Serialize, Deserialize)]
-pub struct Server {
-    pub url: String,
-    pub protocol: String,
-}
-
-#[derive(Debug, Serialize, Deserialize)]
-pub struct Info {
-    pub title: String,
-    pub version: String,
-}
-
-#[derive(Debug, Serialize, Deserialize)]
-pub struct Channel {
-    pub subscribe: Option<Operation>,
-    pub publish: Option<Operation>,
-}
-
-#[derive(Debug, Serialize, Deserialize)]
-pub struct Operation {
-    #[serde(rename = "operationId")]
-    pub operation_id: String,
-    pub summary: String,
-    pub message: Message,
-}
-
-#[derive(Debug, Serialize, Deserialize)]
-pub struct Message {
-    pub payload: Payload,
-}
-
-#[derive(Debug, Serialize, Deserialize)]
-pub struct Payload {
-    #[serde(rename = "type")]
-    pub data_type: String,
-}
-
-pub fn parse_asyncapi_yaml_file(path: &Path) -> Result<AsyncApi, serde_yaml::Error> {
-    let file = File::open(path).unwrap();
-    let reader = BufReader::new(file);
-    let asyncapi_spec = serde_yaml::from_reader(reader)?;
-    Ok(asyncapi_spec)
-}
-
-impl AsyncApi {
-    pub fn get_subscribe_channels(&self) -> Vec<(&String, &Operation)> {
-        self.channels
-            .iter()
-            .filter_map(|(channel_name, channel)| {
-                channel
-                    .subscribe
-                    .as_ref()
-                    .map(|operation| (channel_name, operation))
-            })
-            .collect()
-    }
-    pub fn get_publish_channels(&self) -> Vec<(&String, &Operation)> {
-        self.channels
-            .iter()
-            .filter_map(|(channel_name, channel)| {
-                channel
-                    .publish
-                    .as_ref()
-                    .map(|operation| (channel_name, operation))
-            })
-            .collect()
+use crate::model::{self, AsyncAPI, ReferenceOr};
+use std::{fs, io, path::Path};
+
+pub fn parse_asyncapi_yaml_file(path: &Path) -> Result<model::AsyncAPI, serde_yaml::Error> {
+    let string_content = fs::read_to_string(path).expect("file could not be read");
+    // check if file is yaml or json
+    let parsed = match path.extension() {
+        Some(ext) => match ext.to_str() {
+            Some("yaml") => serde_yaml::from_str::<model::AsyncAPI>(&string_content).unwrap(),
+            Some("json") => serde_json::from_str::<model::AsyncAPI>(&string_content).unwrap(),
+            _ => {
+                panic!("file has no extension");
+            }
+        },
+        None => {
+            panic!("file has no extension");
+        }
+    };
+    Ok(parsed)
+}
+
+pub fn spec_to_pubsub_template_type(spec: AsyncAPI) -> Result<model::PubsubTemplate, io::Error> {
```

Comment:
Das hier wäre die Alternative, eine Transformationsfunktion AsyncAPI Model -> PubsubTemplate Model (spezifisch auf ein Template). Die Idee ist dass wenn wir nur eine beschränkte Anzahl an Templates haben (pubsub, streaming, etc..), dann können wir für jedes Template ein Model (struct) erstellen, das alle benötigten Variablen für das Template schon enthält. Dann kann man dieses Model direkt der gtmpl Funktion übergeben.

Reply:
Ok das ist eine Gute Idee, aber dann würde ich vorschlagen, dass diese structs nicht in den Ordner mit den kopierten kommen, weil ich habe das übersehen und in dieser pr werden im Gegensatz zu vorher noch nicht mehrere publishes und subscribes unterstützt.

Reply:
Stimmt, ich habe jetzt einen neuen Ordner "asyncapi_model" erstellt für die externen structs. Für andere Sachen gibts den template_model Ordner.